### PR TITLE
Add Floyd's Cycle Detection Algorithm

### PR DESCRIPTION
**Pull request by aqib-m31**
### Describe your change:
Add new algorithm for detecting the cycle in a linked list.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by aqib-m31** [comment]
@tianyizheng02 Thanks for help!
All checks passed when I directly added the file via GitHub :octocat:

When I run `pre-commit run --all-files --show-diff-on-failure` on my local machine, it fails and shows different errors.

File: data_structures/linked_list/floyds_cycle_detection.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        >>> linked_list.add_node(1)
+        >>> tuple(linked_list)
+        (1,)
+        """
+        node = self.head
+        while node:
+            yield node.data
+            node = node.next_node
+
+    def add_node(self, data: Any) -> None:
```
[the remark this answers is not in the document]
Comment (tianyizheng02) [reply data_structures/linked_list/floyds_cycle_detection.py:56]: Agreed, please add tests for this function

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

**Review by JeelGajera — APPROVED** [review]
All good 👍

File: data_structures/linked_list/floyds_cycle_detection.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        >>> tuple(linked_list)
+        (1,)
+        """
+        visited = []
+        node = self.head
+        while node:
+            # Avoid infinite loop in there's a cycle
+            if node in visited:
+                return
+            visited.append(node)
```
Comment (tianyizheng02) [inline data_structures/linked_list/floyds_cycle_detection.py:57]:
```suggestion
        visited = set()
        node = self.head
        while node:
            # Avoid infinite loop in there's a cycle
            if node in visited:
                return
            visited.add(node)
```
A set makes more sense for this purpose, IMO

**Review by tianyizheng02 — APPROVED** [review]

File: data_structures/linked_list/floyds_cycle_detection.py (hunk 2 above)
Comment (tianyizheng02) [reply data_structures/linked_list/floyds_cycle_detection.py:57]: Nevermind, forgot that the Nodes aren't hashable

**Comment by tianyizheng02** [comment]
Thanks for your contribution!

**Pull request merged**